File: test/tests/join_one_to_one_test.dart
// @Timeout(Duration(seconds: 10))

import 'package:sangre/sangre.dart';
import 'package:test/test.dart';

void main() {
  test('Join node', () async {
    final ListSource<Map<String, dynamic>> usersDBSource =
        await ListSource<Map<String, dynamic>>();
    for (var i = 0; i < 4; i++) {
      usersDBSource.insertRow({
        'id': i,
        'name': randomString(),
        'friend': 4 - 1 - i,
      });
    }
    expect(usersDBSource.stream.valueOrNull, isNot(null));

    final chain = await JoinOneToOne(
      usersDBSource,
      'friend',
      usersDBSource,
      'id',
    );

    expect(chain.stream.valueOrNull, isNot(null));

    expect(
      chain.stream.value[1]['friend'],
      equals(usersDBSource.stream.value[2]),
    );
  });

  test('Join node & source change', () async {
    // This test has no semantic meaning, but tests source propagation
    final ListSource<Map<String, dynamic>> usersDBSource =
        await ListSource<Map<String, dynamic>>();
    for (var i = 0; i < 4; i++) {
      usersDBSource.insertRow({
        'id': i,
        'name': randomString(),
        'friend': 4 - 1 - i,
      });
    }

    final chain = await JoinOneToOne(
      usersDBSource,
      'friend',
      usersDBSource,
      'id',
    );

    usersDBSource.insertRow({
      'id': 4,
      'name': randomString(),
      'friend': 3,
    });

    await usersDBSource.stream.first;

    expect(
      chain.stream.value[4]['friend'],
      equals(usersDBSource.stream.value[3]),
    );
  });
}

File: test/join_one_to_one.test.ts

import { expect } from 'chai'
import _, { uniqueId } from 'lodash'
import { describe } from 'mocha'
import { JoinOneToOne, ListSource } from '../src/index'

describe("Join one to one", async function () {
    it('Join node', async function () {
        const users =
            await new ListSource<Object>()
        _.times(4, (i) =>
            users.insertRow({
                'id': i,
                'name': uniqueId(),
                'friend': 4 - 1 - i,
            })
        )
        expect(users.value).to.not.be.undefined


        const chain = await new JoinOneToOne(
            users,
            'friend',
            users,
            'id',
        )

        expect(chain.value).to.not.be.undefined
        expect((chain.value as any)[1]['friend']).to.be.equal((users.value as any)[2])
    })


    it('Join node & source change', async function () {
        const users =
            await new ListSource<Object>()
        _.times(4, (i) =>
            users.insertRow({
                'id': i,
                'name': uniqueId(),
                'friend': 4 - 1 - i,
            })
        )
        const chain = await new JoinOneToOne(
            users,
            'friend',
            users,
            'id',
        )

        users.insertRow({
            'id': 4,
            'name': uniqueId(),
            'friend': 3,
        })

        await chain.take(1)

        expect((chain.value as any)[4]['friend']).to.be.equal((users.value as any)[3])
    })
})
